var/spack/repos/builtin/packages/survey/package.py

```diff
@@ -0,0 +1,210 @@
+        # manually or create separate module files, if multiple mpi's
+        # are specified in the build
+        if self.spec.satisfies('+mpich'):
+            env.set('SURVEY_MPI_IMPLEMENTATION', "mpich")
+
+        if self.spec.satisfies('+mvapich'):
+            env.set('SURVEY_MPI_IMPLEMENTATION', "mvapich")
+
+        if self.spec.satisfies('+mvapich2'):
+            env.set('SURVEY_MPI_IMPLEMENTATION', "mvapich2")
+
+        if self.spec.satisfies('+mpt'):
+            env.set('SURVEY_MPI_IMPLEMENTATION', "mpt")
+
+        if self.spec.satisfies('+openmpi'):
+            env.set('SURVEY_MPI_IMPLEMENTATION', "openmpi")
+
+        mpath = '/share/man'
+        env.prepend_path('MANPATH', self.spec.prefix + mpath)
+
+        env.prepend_path('PATH', self.spec['python'].prefix.bin)
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.prefix, self.site_packages_dir))
+        # These are external packages whose site-packages directories are needed
+        # in order to load survey with one module load command.  If these statements
+        # are not present.  The corresponding module files will need to be loaded
+        # These are explicit depends_on packages
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['python'].prefix, self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-pandas'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-python-dateutil'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-setuptools'].prefix,
+                                   self.site_packages_dir))
+        # These are not-explicit depends_on packages
+        # but they must be known in the spec for survey, so it works.
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-numpy'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-pytz'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-six'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-psutil'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-sqlalchemy'].prefix,
+                                   self.site_packages_dir))
+        env.prepend_path('PYTHONPATH',
+                         join_path(self.spec['py-pbr'].prefix,
+                                   self.site_packages_dir))
```

Comment:
Replace this with a for-loop over dependencies, checking whether they extend python and adding them to PYTHONPATH if they do. That allows you to avoid referencing indirect deps.

Reply:
Why are we adding these? If you're extending `python`, then these should already be added.

Reply:
Do you mean something like this?
        for d in dependent_spec.traverse():
            if d.package.extends(self.spec['python']):
                env.prepend_path('PYTHONPATH',
                                             join_path(d.package.prefix, self.site_packages_dir))

Reply:
For some reason they are not.  We can't execute survey without adding these.

Reply:
Have you tried https://spack.readthedocs.io/en/latest/module_file_support.html#autoload-dependencies? Every Python/R/Perl/etc. package in Spack can't be used unless it's dependencies are in PYTHONPATH/R_LIBS/PERL5LIB/etc. We don't want to add this to every package.

Reply:
@adamjstewart Does that mean that every person that wants to build survey needs to set up a modules.yaml file with this in it?   We were hoping people could just build survey and run it out of the box.

Reply:
People should be able to run all Spack packages out of the box. If you're seeing this issue with survey, I'm guessing you're also seeing this issue with every Python/Perl/R package in Spack? We should fix this for all of Spack, not on a package-by-package basis. I opened an issue about autoloading run-time deps (can't find it now). Maybe @alalazo @becker33 @scheibelp would know.

Reply:
@adamjstewart Yes, that would be very helpful!    Is that something that will be in soon, or should I leave the PYTHONPATH settings as is for now?

Reply:
Found it: https://github.com/spack/spack/discussions/8639

Let me bring this up to other admins again.

Reply:
It should really not be necessary to add all these things by hand. The proposed solution

```python
for d in dependent_spec.traverse(): if d.package.extends(self.spec['python']): env.prepend_path('PYTHONPATH', join_path(d.package.prefix, self.site_packages_dir))
```
is almost exactly what the Python package does in `setup_dependent_run_environment`, and that gets added to the module file already.

Are you 100% sure you don't get the PYTHONPATH's in the module file? Can you delete most of setup_run_environment, run spack install [pkg], and share the contents of the module file here?

Reply:
@haampie I will work on taking out some of the setup_run_environment and post here.

Reply:
@haampie You are correct the PYTHONPATH statements are already being generated by spack.  I don't believe this was the case 1 to 1.5 years ago when we were developing this package.py file, unless we just didn't realize it.   I commented out all but the self.spec PYTHONPATH statements and still get these PYTHONPATHs:
```

prepend-path LD_LIBRARY_PATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/lib"
prepend-path LD_LIBRARY_PATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/lib64"
prepend-path PATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/bin"
prepend-path MANPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/share/man"
prepend-path CMAKE_PREFIX_PATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-pandas-1.3.2-tuuxvav3gsowbtypeskwvqai7mtqdxyu/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-bottleneck-1.3.2-juhad5cebjj2ezbs4ddezhwpbahwuyue/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-numpy-1.21.1-7vinivuvvrb3xb7ahcl4u6hlla6zv2wc/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-setuptools-57.4.0-excepexbsoeflig7nst77pi4zihokrok/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-numexpr-2.7.2-2rmtoz2qf4gfh6ent3zaur4iqliwmemy/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-python-dateutil-2.8.2-4hxuu5lzfvnage3wgcpxggytrcepllhf/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-six-1.15.0-5ckqhco4tt4wyhfuul2267eyodb6dqay/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-pytz-2020.1-rduvaznoqx3rrul55geqzlyw5z7qoin2/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-pbr-5.4.3-rfigovp3yrknq5bfvxzuivp4abpnm5cb/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-pip-21.1.2-r6r3a2nq36dgqwqo5fssq4aeyki3fdp6/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-psutil-5.8.0-ik77szm5vmaz7uymnjtanlvg5ysbbyfv/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-sqlalchemy-1.4.20-yco5wdcqu4o3pymjaatttgdydl3j6mvw/lib/python3.8/site-packages"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/py-greenlet-1.1.0-dxri3opvaix34ztkkgr4d3z266cm6kkl/lib/python3.8/site-packages"
setenv SURVEY_MPI_IMPLEMENTATION "openmpi"
prepend-path MANPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/share/man"
prepend-path PATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/python-3.8.11-w6uuvzupzmyntwv36436jfmnozb2bhz5/bin"
prepend-path PYTHONPATH "/home/jgalarowicz/ospack/spack/opt/spack/linux-ubuntu21.04-skylake/gcc-8.4.0/survey-1.0.0-fcj3z2yzmpkwgax5uf7hku6d53fn2ujc/lib/python3.8/site-packages"
```

Reply:
Ok great, it could be this was added at a later stage indeed, I don't recall when. Thanks for testing! Then it'd be good to remove the redundant env modifications

Reply:
@haampie Yes, I'm working on removing all the redundant items that you and Adam have pointed out.  I can't remove the MPI code because we build multiple MPI collectors.  I will check in another mod shortly.   Thanks!